### ISSUE TITLE
Segmentation fault when passing a command

### ISSUE THREAD

**Issue by kantuni** [lydell/run-pty#61]
The tool looks really good, but I'm having a problem running it.

It segfaults every time I try to pass a command to it.
I'm running it on 
- macOS Big Sur (11.7.10)
- Node (version `20.9.0`)
- `npm` (version: `10.1.0`)
- `pnpm` (version: `9.6.0`)
- `bun` (version `1.1.3`)

Here is the exact output:

<img width="627" alt="Screen Shot 2024-07-31 at 1 12 34 AM" src="https://github.com/user-attachments/assets/0c461bc6-fe4f-4139-bdef-dfbbee827e4e">

Please let me know how I can assist with debugging the issue.

**Comment by lydell** [comment]
Hi! My guess is it’s because you use an old version of macOS – I have only tested with the latest macOS. Or it could be Intel vs M1/M2/M3.

Something you can do is trying the underlying `@lydell/node-pty` package directly, which is the `node-pty` package but with prebuilt binaries.

First try `@lydell/node-pty` and see if it works:

```
❯ node
Welcome to Node.js v20.12.1.
Type ".help" for more information.
> require("@lydell/node-pty").spawn("echo", ["test"]).onData(console.log)
{ dispose: [Function: dispose] }
> test
```

Then do the same with `node-pty` and see if that works. I have a feeling the original `node-pty` might work since it builds the binary on your computer at install. If so, there might be something wrong with how I set up the builds for macOS.

Are you running on an Intel Mac then? (Not M1, M2, M3.) I have never tested the x86 macOS build, because I don’t have an old Mac to test on.

Oh, and if `node-pty` works on your computer, as a workaround you could use run-pty 4.1.0 for now (I introduced `@lydell/node-pty` in 5.0.0).

**Comment by kantuni** [comment]
Thanks for a nimble response!

The issue was indeed with `@lydell/node-pty` (and yes, I'm running on an Intel Mac).

Running 
```js
require("@lydell/node-pty").spawn("echo", ["test"]).onData(console.log)
```
results in the above-mentioned segfault. Switching to `4.1.0` did the trick. 

Can we fix the issue in v5 as well? I'm happy to help test it. I'm guessing lots of people still use Intel based MacBooks, and this tool is too good for them to miss out on.

**Comment by lydell** [comment]
You could start by testing if `@lydell/node-pty` _can_ work on your computer at all:

1. Clone https://github.com/lydell/node-pty. The default branch is `prebuild` – that’s the one you want.
2. Run `npm install` to install dependencies and build the native module for your computer.
3. Run `npm run build` to build the JavaScript code.
4. Run `node -e 'require(".").spawn("echo", ["test"]).onData(console.log)'` to verify that it works.

Hopefully that works. If not, I must have messed something up in my fork.

If it does work, then I think there must be something wrong with how I build the binaries in GitHub Actions. You can see how it works here: https://github.com/lydell/node-pty/blob/prebuild/.github/workflows/ci.yml

The tool that builds the binaries is called `node-gyp`. By default, it builds for the OS and arch it runs on. It also has an `--arch` flag to build for another architecture.

I use `macos-14` which runs on an M1 machine. I first build the arm64 binary, then I build again with `--arch=x64`. The `file` command says that the x64 binary is a Mach-O 64-bit x86_64 executable. But it doesn’t seem to work, since it segfaults for you.

Maybe you can try building the x64 binary on macos-13 instead? It runs on an Intel Mac. Maybe it works there.

To try this out:

1. Fork https://github.com/lydell/node-pty.
2. Make a branch and a pull request in your own repo. The workflow should start running. (You can also test without a PR if you edit the workflow triggers, or work directly on the `prebuild` branch.)
3. Comment out the builds for Windows and Linux.
4. Temporarily edit the macOS build to see if you can produce a working macOS Intel binary. Don’t care about the arm64 binary for now, we’ll handle that when we know the x64 binary works.
5. Keep pushing and waiting for CI until you have a working x64 binary, or give up :)

The artifacts (= the binaries) end up in each workflow run. Example: https://github.com/lydell/node-pty/actions/runs/9588703229. Put the `pty.node` and `spawn-helper` files in `./node_modules/@lydell/node-pty-darwin-x64/` to test the files built from GitHub Actions (rather than the ones built locally on your computer).

You can verify that you are actually running the files you think you are running by deleting `pty.node` and `spawn-helper` from there. Then, the test command should output something like:

```
❯ node -e 'require(".").spawn("echo", ["test"]).onData(console.log)'
/Users/you/node-pty/lib/requireBinary.js:12
            throw new Error(optionalDependencies === undefined
            ^

Error: The @lydell/node-pty package could not find the binary package: @lydell/node-pty-darwin-arm64/pty.node
```

**Comment by kantuni** [comment]
Thank you so much for the detailed explanation! 
I've figured out the issue and the [PR](https://github.com/lydell/node-pty/pull/4) with the fix is ready to be reviewed.

**Comment by lydell** [comment]
I’ve published `@lydell/node-pty@1.0.2` with your fix. run-pty depends on `@lydell/node-pty@^1.0.1`, so by re-installing you should get the latest node-pty. Can you try that and confirm it works? Then we can close this issue.

**Comment by kantuni** [comment]
Everything works flawlessly! Thank you for all the support in landing this 💪
